mlir/lib/Dialect/Arith/Transforms/IntRangeOptimizations.cpp

```diff
@@ -214,6 +455,32 @@ struct IntRangeOptimizationsPass
       signalPassFailure();
   }
 };
+
+struct IntRangeNarrowingPass final
+    : arith::impl::ArithIntRangeNarrowingBase<IntRangeNarrowingPass> {
+  using ArithIntRangeNarrowingBase::ArithIntRangeNarrowingBase;
+
+  void runOnOperation() override {
+    Operation *op = getOperation();
+    MLIRContext *ctx = op->getContext();
+    DataFlowSolver solver;
+    solver.load<DeadCodeAnalysis>();
+    solver.load<IntegerRangeAnalysis>();
+    if (failed(solver.initializeAndRun(op)))
+      return signalPassFailure();
+
+    DataFlowListener listener(solver);
+
+    RewritePatternSet patterns(ctx);
+    populateIntRangeNarrowingPatterns(patterns, solver, bitwidthsSupported);
+
+    GreedyRewriteConfig config;
+    config.listener = &listener;
+
+    if (failed(applyPatternsAndFoldGreedily(op, std::move(patterns), config)))
+      signalPassFailure();
```

Comment:
Could we use `walkAndApplyPatterns` here?

Reply:
Actually not:
* We specifically need bottom-up traversal (added a comment)
* We probably want `arith.trunci/exti` folders to run for some final cleanup

Reply:
`walkAndApplyPatterns` uses the post order traversal -- is the lack of folders the issue?

Reply:
The problem is in code
```
func.func @test_add_cmpi() -> i1 {
  ...
  %4 = arith.addi %0, %1 : index
  %5 = arith.cmpi slt, %3, %4 : index
  return %5 : i1
}
```
`arith.cmpi` must be matched before `arith.addi`

The missing folders is not a big problem by itself (and I'm planning to run these patterns in greedy driver anyway in prod), but it makes test code matching more annoying as trunc/ext casts are not removed.

Reply:
OK, I see, so you want this to be visited in the reverse-topo order